src/projection.jl

```diff
@@ -31,114 +68,269 @@ function Base.show(io::IO, project::ProjectTo{T}) where {T}
     end
 end
 
-# fallback (structs)
-function ProjectTo(x::T) where {T}
-    # Generic fallback for structs, recursively make `ProjectTo`s all their fields
+# Structs
+function generic_projectto(x::T; kw...) where {T}
```

Comment:
What else did you try this generic story out on?

I think all the other weird LinearAlgebra types could be treated the same way, I just picked a few as warm-up exercises to see what the machinery could handle. 

My original vote was to go lower-tech, and not build something capable of arbitrary recursive structure.

Reply:
Here's something I feel ought to work, and doesn't. Does this make the right Tangent? Why doesn't the generic method handle that?
```
ProjectTo(x::SymTridiagonal{T}) where {T<:Number} = generic_projectto(x)
function (project::ProjectTo{SymTridiagonal})(dx::AbstractMatrix)
    dv = project.dv(diag(dx))
    ev = project.ev((diag(dx, 1) .+ diag(dx, -1)) ./ 2)
    return SymTridiagonal(dv, ev)
end

(project::ProjectTo{SymTridiagonal})(dx::SymTridiagonal) = project(Tangent{typeof(dx)}(dx)) # maybe?
#=

ss = SymTridiagonal(Symmetric(rand(Int8,5,5)))
pp = ProjectTo(ss)
pp(rand(5,5) .+ im) # works, via necessarily hand-written method

pp(ss) # could be efficient & generic, but how?
Tangent{typeof(ss)}(ss)  # no method matching canonicalize

=#
```

Reply:
Maybe I made this work. I copied one commit to #394 too.

Reply:
> What else did you try this generic story out on?

Nothing, the code that in the PR, is AFAIK only tested on the stuff in the tests.
It was mostly a sketch.

> My original vote was to go lower-tech, and not build something capable of arbitrary recursive structure.

I like this idea.
Seems safe now, and can always generalize later

Reply:
Would have been a lot less work a month or two ago...

Reply:
Are we agreeing or disagreeing.
(It has been a long day)

Reply:
It probably should be possibly to have a recursive definition for that that almost always works.
Since julia does have only a fairly finite number of primative types, like `Symbol` and `Char`.
Plus structs with no fields
And we know we need to just remember any of those, as is.
and we need to apply that recursively.
It does get more confusing to think about recursively though.
Particularly when some structs may have fields that are structs that are made up of some of those, but also of some things that do need to be projected, and so you can't just keep the struct as is.
I think there probably is an answer though that this might already be implementing.
(it's been a looooooooooong day)

Reply:
I don't think recursive things are impossible to make work, as you say it must either terminate with numbers/arrays, or to non-diff things like Symbol. But I worry too that implementing all this struct stuff here is the wrong level, ChainRules has rules for fairly basic things like arrays & numbers. Functors + Optimisers is one machine for recursive structs of stuff with an eye on AD. And there are lots of issues I didn't think of at first, like with weight sharing between different branches of the struct, which get discovered by using it in the wild. Maybe duplicating all of that here, but with different quirks, is a terrible idea, especially in a package that aims to be a stable core thing for simple shared meanings. What's the most complicated Tangent actually in use in the wild? I haven't really tried to follow that.

My vote for projection was to kick the can down the road and get the basic projection for arrays & numbers up & running & in use in the wild, see what breaks. I wanted to do this in Zygote precisely because it could be used right away, and by the time we throw Zygote away we'll know how to design a better one. But nobody read the PR, I don't think, or thought about its strengths & weaknesses.

So today mostly I'm just disappointed that we land up discussing the first-order issue after getting literally hundreds of comments deep into the weeds, and being stressed about getting it right for v1. And I despair at finding a better process for doing this, somehow we had about 10 implementations (including several of mine) and there were so many details to discuss and so many lists of what details not to forget to discuss that somehow actually figuring out the scope of the whole thing was apparently left up to accident... this can't be a great way to use everyone's time.

Reply:
Also, I really need a holiday.

The present state isn't so bad. The fancy recursive stuff is only 10-20 lines, and doesn't get called on unknown arbitrary structs... maybe it never does, maybe it gets removed later. For now it's used on some random LinearAlgebra cases nobody really needs.